src/chk/chk_iv.c

```diff
@@ -0,0 +1,236 @@
+/**
+ * (C) Copyright 2022 Intel Corporation.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause-Patent
+ */
+
+#define D_LOGFAC	DD_FAC(chk)
+
+#include <gurt/list.h>
+#include <gurt/debug.h>
+#include <cart/iv.h>
+#include <daos_srv/iv.h>
+#include <daos_srv/daos_engine.h>
+
+#include "chk_internal.h"
+
+static int
+chk_leader_notify(uint64_t gen, d_rank_t rank, uint32_t phase, uint32_t status)
+{
+	/* XXX: to be implemented in subsequent patch. */
+	return 0;
+}
+
+static int
+chk_engine_notify(uint64_t gen, uuid_t uuid, d_rank_t rank, uint32_t phase,
+		  uint32_t status, bool remove_pool)
+{
+	/* XXX: to be implemented in subsequent patch. */
+	return 0;
+}
+
+static int
+chk_iv_alloc_internal(d_sg_list_t *sgl)
+{
+	int	rc;
+
+	rc = d_sgl_init(sgl, 1);
+	if (rc != 0)
+		return rc;
+
+	D_ALLOC(sgl->sg_iovs[0].iov_buf, sizeof(struct chk_iv));
+	if (sgl->sg_iovs[0].iov_buf == NULL) {
+		d_sgl_fini(sgl, true);
+		return -DER_NOMEM;
+	}
+
+	sgl->sg_iovs[0].iov_buf_len = sizeof(struct chk_iv);
+	sgl->sg_iovs[0].iov_len = sizeof(struct chk_iv);
+
+	return 0;
+}
+
+static int
+chk_iv_ent_init(struct ds_iv_key *iv_key, void *data, struct ds_iv_entry *entry)
+{
+	int	rc;
+
+	rc = chk_iv_alloc_internal(&entry->iv_value);
+	if (rc == 0) {
+		entry->iv_key.class_id = iv_key->class_id;
+		entry->iv_key.rank = iv_key->rank;
+	}
+
+	return rc;
+}
+
+static int
+chk_iv_ent_get(struct ds_iv_entry *entry, void **priv)
+{
+	return 0;
+}
+
+static int
+chk_iv_ent_put(struct ds_iv_entry *entry, void **priv)
+{
+	return 0;
+}
+
+static int
+chk_iv_ent_destroy(d_sg_list_t *sgl)
+{
+	d_sgl_fini(sgl, true);
+
+	return 0;
+}
+
+static int
+chk_iv_ent_fetch(struct ds_iv_entry *entry, struct ds_iv_key *key, d_sg_list_t *dst, void **priv)
+{
+	D_ASSERT(0);
+
+	return 0;
+}
+
+/* Update the chk pool svc lists and status from engine to leader. */
+static int
+chk_iv_ent_update(struct ds_iv_entry *entry, struct ds_iv_key *key,
+		  d_sg_list_t *src, void **priv)
+{
+	struct chk_iv	*dst_iv = entry->iv_value.sg_iovs[0].iov_buf;
+	struct chk_iv	*src_iv = src->sg_iovs[0].iov_buf;
+	int		 rc = 0;
+
+	/*
+	 * When check leader (IV master) tirgger chk_iv_update, it will set ci_to_leader as 0,
+	 * then chk_iv_ent_update() for any IV message from leader will get -DER_IVCB_FORWARD.
+	 */
+	if (!src_iv->ci_to_leader)
+		D_GOTO(out, rc = -DER_IVCB_FORWARD);
```

Comment:
it looks confuse to me, as the IV leader/master call chk_iv_update() with ci_to_leader == 0, on local node of the leader, cart will call chk_iv_ent_update() for the leader, here it returns DER_IVCB_FORWARD means ask cart to forward to leader, but itself is the leader already, seems does not work? may confirm with @wangdi1

Reply:
In DAOS check infrastructure, there is unique check leader and multiple check engines. On one server rank, there can be at most one check engine. But the check leader and check engine can be on the same rank. There is requirement to send message via IV from the check leader to the check engine on the same rank. I prefer to handle such case with the same logic as the case that check leader and check engine on different ranks. Does IV support such use case?

Reply:
check leader and check engine can be different ULT on same daos_engine (same rank)?
as they operate on same IV class with same set of call-backs, seems cannot work, or say I don't know how does it work that make cart differentiate the check leader and check engine's role?

Reply:
> check leader and check engine can be different ULT on same daos_engine (same rank)?
Yes, it is possible.

> as they operate on same IV class with same set of call-backs, seems cannot work, or say I don't know how does it work that make cart differentiate the check leader and check engine's role?

You mean crt does not support the message from one ULT to another ULT on the same rank?

Reply:
> You mean crt does not support the message from one ULT to another ULT on the same rank?

cart RPC can do that (in ULT_1 send a RPC to itself, and the RPC received/service in another ULT), no problem.
But for IV message passing, if ULT_1/_2 use the same IV class and the same set of IV call-backs, looks cart cannot differentiate them - like who is the IV's leader and who is participator.